### PR TITLE
GH-512 As a developer I need easy fix for all missing dependencies

### LINKED ISSUE #512: As a developer I need easy fix for all missing dependencies
While there are few ways to install multiple npms into IDE (target platform file, multi-quikfix, dialog in library manager) it still may be inconvenient to use them, especially when setting up new workspace with large existing projects.

To ease fixing multiple missing dependencies it would be good to provide easily accessible `try-to-install-all-missing-dependnecies` functionality. This _überfix_ can be provided as separate button in the IDE, to allow easy access for the user.

### NOTE
When working on this look into existing quick fix. The existing one could be refactored to invoke `NpmManager` once for all missing dependencies. The same behaviour is desired for the requested feature.

### PR DESCRIPTION
**Pull request by yoosiba**
resolves #512

### CODE REVIEW

File: plugins/org.eclipse.n4js.ui/plugin.xml [hunk 1]
```diff
@@ -71,22 +71,6 @@ Contributors:
             id="org.eclipse.n4js.ui.export">
       </category>
 
-<!-- Disabled NFAR Wizard -->
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ui/plugin.xml]: good

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/external/ExternalLibraraiesActionsHelper.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+import org.eclipse.n4js.ui.preferences.external.MaintenanceActionsButtonListener;
+import org.eclipse.n4js.utils.StatusHelper;
+import org.eclipse.n4js.utils.io.FileDeleter;
+
+import com.google.inject.Inject;
+
+/**
+ * Similar to {@link MaintenanceActionsButtonListener} actions, but dedicated for different UI.
+ */
+public class ExternalLibraraiesActionsHelper {
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/external/ExternalLibraraiesActionsHelper.java]:
```
ExternalLibraraiesActionsHelper
              ^
```

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/handler/DependneciesDialog.java [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+import com.google.common.base.Strings;
+
+/**
+ * UI for the {@link LibrariesFixHandler}, which will use this class for visualizing its long running operation. Since
+ * in some (user data depended) scenarios processing has to be suspended in order to get users input, instance of this
+ * class will hold reference to the caller. It is assumed that, if needed, the caller will {@link #wait()} for this
+ * instance to get user input, and this instance will {@link #notify()} caller when input is obtained.
+ */
+public class DependneciesDialog extends ProgressMonitorDialog {
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/handler/DependneciesDialog.java]:
```
DependneciesDialog
      ^
```

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/handler/LibrariesFixHandler.java [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+				@Override
+				public void run(IProgressMonitor monitor) {
+					IStatus status = setupWorkspaceDependnecies(monitor, dependneciesDialog);
+					if (!status.isOK())
+						userLogger.logError(status);
+					monitor.done();
+				}
+
+			};
+			dependneciesDialog.run(true, true, iRunnableWithProgress);
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/handler/LibrariesFixHandler.java]:
```
dependneciesDialog.run
      ^
```

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/handler/ProjectDependneciesHelper.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+import org.eclipse.n4js.projectModel.IN4JSCore;
+import org.eclipse.n4js.projectModel.IN4JSProject;
+import org.eclipse.n4js.ui.internal.EclipseBasedN4JSWorkspace;
+
+/**
+ * Allows to access data in the manifests of the provided {@link IN4JSProject}s. Unlike {@link N4JSModel} will return
+ * data that describes unknown or missing projects. Unlike {@link ManifestValuesParsingUtil} it allows to read manifest
+ * of known projects, and not arbitrary {@code .n4mf} file.
+ */
+class ProjectDependneciesHelper {
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/handler/ProjectDependneciesHelper.java]:
```
ProjectDependneciesHelper
             ^
```

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/handler/ProjectDependneciesHelper.java [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+	 * workspace based data wins, e.g.
+	 *
+	 * <pre>
+	 *  <ul>
+	 *   <li> platform files requests "express" but no project depends on "express" then express is installed (in latest version)</li>
+	 *   <li> platform files requests "express@2.0.0" but workspace project depends on "express@1.0.0"  then express is installed in version "2.0.0"</li>
+	 *  <ul>
+	 * </pre>
+	 */
+	public Map<String, String> calculateDependneciesToInstall(File selectedN4TP) {
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/handler/ProjectDependneciesHelper.java]: take a guess

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/handler/ProjectDependneciesHelper.java [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+		}
+		return new HashMap<>();
+	}
+
+	/**
+	 * Calculates missing dependencies based on current workspace and library manager state.
+	 *
+	 * Names of the dependencies are not validated. Also there is no version resolution, last one wins.
+	 */
+	private void updateMissingDependneciesMap(Map<String, String> versionedPackages) {
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/handler/ProjectDependneciesHelper.java]: STRG-F

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/external/NpmManager.java [hunk 8]
```diff
@@ -174,7 +174,7 @@ public IStatus installDependencies(final Map<String, String> versionedNPMs, fina
 	 * @param monitor
 	 *            the monitor for the blocking install process.
 	 * @param triggerCleanbuild
-	 *            iff true, a clean build is triggered on all affected workspace projects.
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/external/NpmManager.java]: are you aware that: "iff" = "if and only if". In this case here "iff" seems more precise.

Comment (yoosiba) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/external/NpmManager.java]: yes but down the code there is `if (Platform.isRunning())`, then there is dispatch to sub type implementation where one implementation does nothing, while other one can exit early in certain conditions.

So.. it is actually less `iff` and more `iffy`.

Comment (mmews-n4) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/external/NpmManager.java]: 👍

**Pull request merged**